packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java

```diff
@@ -470,6 +511,49 @@ public void copyAsync(Map<String, Object> options, Promise promise) {
     }
   }
 
+  private void transformFilesFromSAF(DocumentFile documentFile, File outputDir, boolean copy) throws IOException {
+    if (!documentFile.exists()) {
+      return;
+    }
+
+    if (!outputDir.exists()) {
+      if (!outputDir.mkdirs()) {
+        throw new IOException("Couldn't create folder in output dir.");
+      }
+    }
+
+    if (documentFile.isDirectory()) {
+      for (DocumentFile file : documentFile.listFiles()) {
+        if (documentFile.getName() == null) {
+          continue;
+        }
+
+        transformFilesFromSAF(file, new File(outputDir, documentFile.getName()), copy);
```

Comment:
is it intentional that if `documentFile` is a directory, all the contents will be copied directly into `outputDir`, rather than to a directory inside of `outputDir`? it seems like if there are nested directories in `documentFile` this will flatten them.

Reply:
It will create a new folder in the outputDir for each directory in the input path (see a new file from the outputDir and document name). So the hierarchy should be preserved.